### DISCUSSION TITLE
雑多なトーク

### DISCUSSION THREAD

**Discussion by hayashi-ay** [katataku/minishell#6] [General]
どのトピックで話していいか迷ったらとりあえずここで。

**Comment by hayashi-ay** [comment]
Kickoffまでに何か準備することあります？

**Reply by katataku** [reply]
考えてなかった・・・
なんか事前に同じ資料読みたいですね。
今読んでるものあれば教えてください。僕も読んでおきます！

**Reply by hayashi-ay** [reply]
なんも読んでないです！

**Reply by hayashi-ay** [reply]
強いていうならUNIXシステムコール・プログラミングの6章ですかね〜？ちょっとしたシェルを作ってます。

**Reply by katataku** [reply]
それだ

**Reply by katataku** [reply]
機能洗い出しに使えそうなinputはあるといいなー。
manはリファレンス的に読みたいから通読したくないなー。

**Reply by hayashi-ay** [reply]
じゃあ宿題は、「UNIXシステムコール・プログラミングの6章」を読んで機能を考えておくにしましょー。

**Reply by katataku** [reply]
子プロセスでpipeする件が腹落ちしてない。
「（読み手にパイプを作成するように依頼した場合）invokeを呼ぶ直前にpipeを呼び出すことができ、それによってパイプラインを任意の長さにすることができる。」
？？？

**Reply by katataku** [reply]
https://github.com/katataku/minishell/wiki/設計
腹落ち度が低いから、整理のために書いてみた

**Comment by hayashi-ay** [comment]
Kickoff何します？

**Reply by katataku** [reply]
しばらくの目標は「個人ワークができるようにタスクを切る」ですよねー。
思いつきですが、
- 大枠の方針
- それぞれのパートのIF決め
をする

**Reply by hayashi-ay** [reply]
あと開発のための環境作りもしたいです。でもやりながらでもいいかも？

**Reply by katataku** [reply]
環境づくりってどんなイメージです？

**Reply by hayashi-ay** [reply]
やるかどうかも含めてですがGithub Actionsとかだったりですかね？

**Reply by hayashi-ay** [reply]
自動テスト的なのを導入するなら。

**Reply by hayashi-ay** [reply]
でもめんどくさい気持ちもあります。

**Reply by katataku** [reply]
Github Actionsかどうかは置いといて、shell scriptでテストコード書きたい欲があります！
でも、どうせやるんだったらGithub Actions熱あります😃

**Reply by hayashi-ay** [reply]
統合テストはShellがいいかもですね〜。単体テストはGoogleテストでいいかなーと思ってます。

**Reply by hayashi-ay** [reply]
序盤の方に簡易的なプログラムと統合テストの骨格+Github Actionsは整備しておきたいかも。。

**Reply by katataku** [reply]
それです！

**Reply by katataku** [reply]
内容空っぽでも「とりあえずテストが動いている状態」までは持って行ってから本格的な非同期作業に入りたいですよね。
ベースラインとしてののテストがあると非同期作業しやすそう

**Reply by hayashi-ay** [reply]
賛成です。

**Reply by hayashi-ay** [reply]
「プロンプトを表示して単純なコマンドを実行する」ところまでやるのがちょうどよいかも。

**Comment by hayashi-ay** [comment]
冬のレビューキャンペーンってどうされます？参加するか迷ってるのですが、minishellの進行にも影響するなーと思っていて。

**Reply by katataku** [reply]
僕はやらないつもりですー。
全然参加しちゃってください！多少遅れても大丈夫ですよ

**Reply by hayashi-ay** [reply]
ありがとうございます。片岡さんやらないならぼくもやらないでおこうかな〜。
課題公開されてから考えようと思います〜。

**Comment by hayashi-ay** [comment]
あ、bashのpdf化失敗してますね。
https://github.com/katataku/minishell/blob/main/docs/man/bash.pdf

**Reply by hayashi-ay** [reply]
```
groff -Tps -mandoc -c /usr/share/man/man1/bash.1 | ps2pdf  -> bash.pdf
```

このコマンドでやったのですが、デカすぎて何らかのエラーになってるとかですかね？

**Reply by hayashi-ay** [reply]
できたかも。

**Comment by hayashi-ay** [comment]
多段パイプだいたいできてきました〜。

**Comment by hayashi-ay** [comment]
シグナル周りも何かしらしないといけなさそうな雰囲気

**Reply by katataku** [reply]
コマンド実行中にシグナルを受け取った時の挙動。
みたいな感じですか？

**Reply by hayashi-ay** [reply]
適当に言いました〜、たぶんそんな感じだったり、bashのプロンプト受付中とかのシグナル入力ですね〜。

**Comment by hayashi-ay** [comment]
オシャレなウェルカムメッセージも作りたいかも

**Reply by katataku** [reply]
```
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM""7.      !77"WMMMMMMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMMMMMM@7` ..JxOtttrttrttO&(.. ?YMMMMMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMMMMM"` .JtttttttttttrtttttrttttO-. ?HMMMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMMMM@^ .ztrtttrtrtrtrtttrtrtttrttttttt&..TMMMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMMMM@!.(ttttrttrtttrtttrtrtrrrtrttrtrtrttrO. TMMMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMM#! (tttttrttrttrtttrtv!`     _?ttrtttrttttO..WMMMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMMMD .tttttrtttrttttrtZ!            ?ttrtttttttro.?MMMMMMMMMMMM
MMMMMMMMMMMMMMMMMMM3 (tttrtttrtttrtrtZ!           ..(.,OttrtrtttttO,,MMMMMMMMMMM
MMMMMMMMMMMMMMMMMM^ Jtrttrttrttrtttrc           .C` .XZJtttrtrtrtttt-.MMMMMMMMMM
MMMMMMMMMMMMMMMMM$ (rttrttrtttrtttr>           Jun-..uu0Otttttrttrrtt-,MMMMMMMMM
MMMMMMMMMMMMMMMMF .rtttttttrtttrtrc           .ZuuZuuZZkjrtrtttrc`   ?.JMMMMMMMM
MMMMMMMMMMMMMMMM} ztrtrtrtttrttttr`          `duZuZuZuu0jtttrtr>  .-+.~.MMMMMMMM
MMMMMMMMMMMMMMM# .rttttrtrtttrtttI            wuZuuZuZZ\zttrttI  ,` (ZI JMMMMMMM
MMMMMMMMMMMMMMMF (ttrtttttrttrtrtI            (ZuZuZuZC.rttttr: .ZXwZuu..MMMMMMM
MMMMMMMMMMMMMMMF jttrtrttrttrttrtr-            (TZZZV^ jtrtrtv  dZuuuuX!.MMMMMMM
MMMMMMMMMMMMMMM{ jrttttrtttrtv<`.`_`       `          ,rttttr}  wZuZZZI~.MMMMMMM
MMMMMMMMMMMMMMM{ jttrtttrttrc``````...              .Jrtrtttr~  (ZuuZV(`.MMMMMMM
MMMMMMMMMMMMMMM{ zttrtrttrttl.`.`.````.`..         `.``. _?1t.   .77^ c JMMMMMMM
MMMMMMMMMMMMMM# .trttrttttrtt-.`.`.``.``.`.`.`.``.``.`.`..`. `       J!.MMMMMMMM
MMMMMMMMMMMMMMF ,ttrtttrtttrtri.``.`.``.```.``.`.``.```.`````..     (> dMMMMMMMM
MMMMMMMMMMMMMM} ztrtrttrtrtttttO-.`.`.``.`.``.``.`.`.`.``..````.` .zv .MMMMMMMMM
MMMMMMMMMMMMMF .ttttrtttrttrttrttr&..`.``.`.`.`.```.``.``.`..`.`.`..` !   .....?
MMMMMMMMMMMM#`.ttrttrttrttttrttttttttO-. ``.``.`.`.``.`.````.``.``.+ztttttttt7 .
MMMMMMMMMMMM^ jrttrttrtttrtttrtrtttttrtrrO+-.`.``.`.``.`.`.``.``.``.zttrttt7 .dM
MMMMMMMMMMM^ Jrttrttrtttrtrtttrtrrv7<!_............ .``.`.`.``.``.`..OtrO> .MMMM
MMMMMMMMMMt Jttrttrtrttrtttrttrv<_................_1rrtOzz-.`.`.``.``(Z! .MMMMMM
MMMMMMMMMF ,tttrttrttrttttrtZ>_.....................(ttttttto. `.``.`. .MMMMMMMM
MMMMMMMMF .rtrttttrtttrttrt>_.......`..`..`..`..`....(rttrttttt&...``.  MMMMMMMM
MMMMMMM@ .rtttrrttrtrttrtt:......`....................jtrttrtttrv!  .`  JMMMMMMM
MMMMMM#`.rttrttrttrttttttr<..........`...`..`..`..`...(rttttrtv!.(NJ,. .MMMMMMMM
MMMMMM' zttrtttrttrtrtrttrl..`..`..`..................-Ottrt7`.dMMMMMMMMMMMMMMMM
MMMMMF (rtttrttrttrttrttrtI...........`...`.....`..`...jtr2`.dMMMMMMMMMMMMMMMMMM
MMMMF .ttrtttrtttrttrtttttt_..`..........`...`.........jt> jMMMMMMMMMMMMMMMMMMMM
MMMM! Otttrtttrtttrtttrttrt:....`..`..`..........`..`..jZ (MMMMMMMMMMMMMMMMMMMMM
MMMF (rttrttrttrttrtrttrtttl..............`...`........jI dMMMMMMMMMMMMMMMMMMMMM
MMM .rttrttrtttrttrtZ?ztrttI.`..`..`...`.....`....`....jI (MMMMMMMMMMMMMMMMMMMMM
MMF (tttttttrttrttZ!..._?1tI_.........`....`........`..jI (MMMMMMMMMMMMMMMMMMMMM
MM .rttrtrtttrttr>_............`..`............`.......jO (MMMMMMMMMMMMMMMMMMMMM
MF (trttrtrtttrZ!...................`...`...`....`...`.jt (MMMMMMMMMMMMMMMMMMMMM
M\ zttrttttrtrZ~..........`..`...`........`.......`....zt.(MMMMMMMMMMMMMMMMMMMMM
M .rtttrttrttO!.....`..`.............`........`........zt.(MMMMMMMMMMMMMMMMMMMMM
```

**Reply by hayashi-ay** [reply]
草ww

**Reply by hayashi-ay** [reply]
変換ツール的なのあるんですか？

**Reply by katataku** [reply]
https://tool-taro.com/image_to_ascii/
これで変換しましたー

**Comment by hayashi-ay** [comment]
マージしたら自動的にブランチが削除されるようにしたいかも。

**Reply by hayashi-ay** [reply]
https://docs.github.com/ja/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-the-automatic-deletion-of-branches

**Reply by katataku** [reply]
対応しましたー

**Reply by katataku** [reply]
マージしたと、issueって自分でクローズしないといけないんですか？

**Reply by hayashi-ay** [reply]
そんなこともないです。
コミット名にfix #(Issue番号)みたいに入っているか、PRの関連Issueに紐付けられていればマージ時にCloseされます。

**Reply by hayashi-ay** [reply]
あとデフォルトのマージをsquashマージにしたいかもです。ぼくSetting権限なく。

**Reply by katataku** [reply]
すいません、間違えました☠️

squashマージだけ有効にしました！
これで間違えなくなるはず！！

**Comment by hayashi-ay** [comment]
コメントのresolveはコメントした側がしたいかも。

**Reply by katataku** [reply]
すいません！そうしましょう

**Comment by hayashi-ay** [comment]
42ヘッダーは今の時点からつけててもいいと思ったり。

**Reply by katataku** [reply]
こだわりないので賛成です。
ただ、ヘッダのタイムスタンプだけ更新差分あるファイルはコミットに含めない方が嬉しいですー

**Comment by hayashi-ay** [comment]
pipexいい感じなんで良きタイミングで https://github.com/katataku/42_pipex_pair にPRあげますね

**Reply by hayashi-ay** [reply]
修正差分がありすぎてかなり見づらいかも。。
https://github.com/katataku/42_pipex_pair/pull/1

**Reply by katataku** [reply]
読むのつらかったorz
ので、木曜日に概要教えてくださーい。

**Reply by hayashi-ay** [reply]
OKです。肝心なところはここだけです〜。
- [typedefs.h](https://github.com/katataku/42_pipex_pair/blob/7022c90f5c504082087de606211f9b1cd5731ea4/works/includes/typedefs.h)
- [exec_pipex.c](https://github.com/katataku/42_pipex_pair/blob/7022c90f5c504082087de606211f9b1cd5731ea4/works/srcs/shared/exec_pipex.c)
- [parser.c](https://github.com/katataku/42_pipex_pair/blob/7022c90f5c504082087de606211f9b1cd5731ea4/works/srcs/pipex_bonus/parser.c)

**Comment by hayashi-ay** [comment]
bashではなくてIEEEの規格を参照した方が良いかもという説。
https://pubs.opengroup.org/onlinepubs/9699919799/utilities/cd.html

**Reply by hayashi-ay** [reply]
IEEEの規格がPOSIXということかな？

**Reply by hayashi-ay** [reply]
> 課題で要求されてるか微妙なラインのbash zshが勝手にやってる系エッジケースはIEEE規格でshellに要求されてないってディフェンスでいけそうな気がしてきた 

のりさんのtimesより抜粋

**Reply by katataku** [reply]
問題文が、
If you have any doubt about a requirement, take bash as a reference.
なので、bashリファレンスを見た上でIEEEも参考にするがいいかなと思いましたー。

とはいえ、大方針として、「課題で要求されてるか微妙なラインのbash zshが勝手にやってる系エッジケース」は「Anything that is not asked is not required.」でディフェンスしましょ〜

**Comment by hayashi-ay** [comment]
ヒアドキュメントも基本的なところはできた！

**Comment by hayashi-ay** [comment]
実行部分の実装、だいたい終わりました〜。

**Comment by hayashi-ay** [comment]
historyの記録はどこだろう？実はlexer？

**Reply by hayashi-ay** [reply]
mainから別で呼ぶ感じかな？

**Reply by katataku** [reply]
readlineにhistory機能あるので、それで賄えるかもって思ってます！

**Comment by hayashi-ay** [comment]
エスケープは対応する？
```
Not interpret unclosed quotes or special characters which are not required by the
subject such as \ (backslash) or ; (semicolon).
```

このバックスラッシュは複数行についての話かもしれないと思い始めました。

**Reply by katataku** [reply]
対応しないと思ってました😣
エスケープ記号としての \も`special characters which are not required by the subject`だと思います

**Reply by hayashi-ay** [reply]
そうしましょー。

**Comment by katataku** [comment]
今回の課題で、シグナルって何に使うんでしょうか？

**Reply by hayashi-ay** [reply]
終了ステータスを合わせるくらいしか思いついてないです。 128 + n

**Reply by hayashi-ay** [reply]
課題文の以下のところですね〜。

```
• Handle ctrl-C, ctrl-D and ctrl-\ which should behave like in bash.
• In interactive mode:
    ◦ ctrl-C displays a new prompt on a new line.
    ◦ ctrl-D exits the shell.
    ◦ ctrl-\ does nothing
```

**Comment by hayashi-ay** [comment]
`exit | exit`はexitしない。

builtin関数はパイプを使う場合は子プロセスで通常は親プロセスで実行されるっぽい。

**Reply by katataku** [reply]
知見！

**Reply by hayashi-ay** [reply]
zshはパイプでもbuiltinを親プロセスで実行してそう

**Comment by hayashi-ay** [comment]
学習という観点でいくと敢えて役割分担を途中で入れ替えるのもありかもしれない。

**Reply by katataku** [reply]
面白いですね！

issueの途中で入れ替えると気持ち悪いので、
担当issueの領域を、バラバラになるようにアサインするの楽しそうです！

**Reply by hayashi-ay** [reply]
やっちゃいますか。たぶん今までの感じを踏襲すると以下みたいになると思うんですけど次のチケット割り振りで逆にしちゃいますか。
 - lexer, parser(片岡)
- 実行部分, ヒアドキュメント(林)

**Reply by hayashi-ay** [reply]
so_longでしげさんに書いてもらってた部分を自分で再実装しているのですが、意外とコード書かないと分からないこととかがあることに気づいたのがきっかけです。

**Reply by katataku** [reply]
そうしましょう！

ヒアドキュメントあたりとか、そうなりそうな感じがしてたのでめっちゃ楽しみ！

**Comment by katataku** [comment]
このエラーってどのタイミングで返すのがスマートですかね？
構文解析の中でやる環境変数の展開のタイミングですかね。

```bash
bash-3.2$ echo ${HO"ME"}
bash: ${HO"ME"}: bad substitution
```

**Reply by hayashi-ay** [reply]
変数名がおかしい場合ってことですよね。
`${1HOME}`とか数字から始まる場合もエラーになりますね。

**Reply by hayashi-ay** [reply]
lexerよりかは変数の展開するやつの責務ですかね〜？

**Comment by hayashi-ay** [comment]
bashってEOF入力されると終了するんですね。

**Reply by hayashi-ay** [reply]
課題要件にもあった

**Comment by hayashi-ay** [comment]
今週のチケットは終わりました〜。分量かなり少なかったんで、Ctrl + CとCtrl + \の対応もやっちゃいます。

**Reply by katataku** [reply]
イケメン！

**Comment by hayashi-ay** [comment]
モジュールに属さないソースフアイルはどこに置きましょうかね？signal周りの処理はset_signal.cみたいなファイルに切り出せそうですが、どこに置こうかなと。srcs/直下ですかね。

**Reply by katataku** [reply]
signalの例だけ聞くと、signalモジュール作っちゃってもいいかなと思いました。

それ以外でモジュール化難しければsrcs直下に賛成ですー

**Reply by hayashi-ay** [reply]
一旦雑にmain.cに入れてあります。singalの扱いやモジュールに切り出すのが難しいファイルの置き場所は来週話したいです〜

**Comment by hayashi-ay** [comment]
次のタスク割り振りの時にプランニングポーカー的なのやってみてもいいかも。
ルーレット決めの前にタスクをなるべく均等に分けるのが目的です。ちゃんと実測値まで図ってみたいなのは考えてないです。

**Reply by katataku** [reply]
いいですねーやってみましょー！

**Comment by katataku** [comment]
思いのほかシンプルにパイプ対応ができた！
（メモリは余分に確保してますが。）

今週末に結合やってみますねー

**Comment by hayashi-ay** [comment]
indexの命名は好みが分かれてそう。
idx_wordよりwi派

**Reply by katataku** [reply]
合わせます！

**Reply by hayashi-ay** [reply]
ありがとうございます。normで対応できない命名規則などは軽くdocsにまとめてもいいかも？

**Comment by katataku** [comment]
気が向いたので、（優先度無視して）リダイレクト対応に手をつけてしまいましたー

**Reply by hayashi-ay** [reply]
気が向いて全部実装終わらせてもらっても良いですよ〜

**Comment by hayashi-ay** [comment]
captureする代わりに出力先を変えることもできるかも？
https://blog.emattsan.org/entry/20120613/1339584605

**Comment by hayashi-ay** [comment]
タブ補完もreadlineが勝手にやってくれてるのか。すごい。

**Comment by hayashi-ay** [comment]
処理効率などを考えると環境変数はコピーオンライトが良さそう。ただ実装がちょっとめんどくさいので、対応はしないのが良さそう。

**Comment by katataku** [comment]
./minishellプログラムのに引数が渡された場合の挙動どうしますか？
本家ならshellscriptとして解釈して実行するぽいです。

対応せずに「引数があればエラー」でもいいかもと思ってます。

**Reply by hayashi-ay** [reply]
引数は無視して、毎回対話モードで起動する感じでいいかなと思ってます。

**Comment by katataku** [comment]
すいません〜間違えてmainにmergeしちゃったんでrevertしましたm(_ _)m

**Comment by hayashi-ay** [comment]
なんかこのdiscussion長くなっちゃったんで別で立てなおしてもいいかもですね。
